Initial setup

srueg opened the pull request:
Set up a Crossplane stack for cloudscale.ch

Closes #2

Commit:
Initial setup of types

Commit:
Set up managed resource controller for S3Bucket

Commit:
Streamline Makefile and Dockerfile

Commit:
Write README

Commit:
Set up claim and class scheduling, add samples

Commit:
Implement Cloudscale API

To create Object Users.

Commit:
Implement S3 API

To create and delete buckets.

Commit:
Rename project

Rename CRD group to cloudscale.crossplane.io and move repo to
github.com/vshn/stack-cloudscale.

hasheddan commented on api/v1alpha1/s3bucket_types.go:
```diff
@@ -0,0 +1,132 @@
+	// first instance of the string '%s' will be replaced with the Kubernetes
+	// UID of this S3Bucket. Omit this field to use the UID alone as the name.
+	// +optional
+	NameFormat string `json:"nameFormat,omitempty"`
```
Not 100% required at this stage by any means, but it is standard to make optional fields pointer values (i.e. `*string` here). @muvaf wrote up some great documentation on this that directly references the k8s documentation as well https://github.com/crossplaneio/crossplane/blob/master/design/one-pager-managed-resource-api-design.md#pointer-types-and-markers

hasheddan commented on api/v1alpha1/s3bucket_types.go:
```diff
@@ -0,0 +1,132 @@
+// provisioned S3Bucket.
+type S3BucketClassSpecTemplate struct {
+	runtimev1alpha1.ClassSpecTemplate `json:",inline"`
+	S3BucketParameters                `json:",inline"`
```
Another "newer" pattern we have begun to standardize around is call the embedded parameters struct `forProvider`. Once again not required by any means, but it may be beneficial in standardizing around other stacks patterns. Then, similarly, we have an `S3BucketObservation` struct that we embed in the `status` as `atProvider`.

Example Ref: https://github.com/crossplaneio/stack-aws/blob/81f612019be228db4c093b6879f4485d61ef9d08/apis/database/v1beta1/rdsinstance_types.go#L818

Docs Ref: https://github.com/crossplaneio/crossplane/blob/master/design/one-pager-managed-resource-api-design.md#owner-based-struct-tree

hasheddan commented on api/v1alpha1/s3bucket_types.go:
```diff
@@ -0,0 +1,132 @@
+//   3. NameFormat = "foo", BucketName = "foo"
+//   4. NameFormat = "foo-%s", BucketName = "foo-test-uid"
+//   5. NameFormat = "foo-%s-bar-%s", BucketName = "foo-test-uid-bar-%!s(MISSING)"
+func (b *S3Bucket) GetBucketName() string {
```
Have added some comments below that relate to potentially removing this.

hasheddan commented on controllers/s3/s3bucket.go:
```diff
@@ -0,0 +1,236 @@
+	}
+	log.Info("Create", "bucket", bucket.Name)
+
+	objectUser, err := e.s3Client.CreateOrUpdateBucket(ctx, bucket.Status.ObjectUserID, bucket.GetBucketName(), bucket.Spec.Tags)
```
We have begun to standardize around using an annotation to represent the external name of a resource. Because it appears that this provider allows you to specify external name before creation, the annotation should be set on the managed resource using `meta.SetExternalName()` and retrieved via `meta.GetExternalName()`.

Doc Ref: https://github.com/crossplaneio/crossplane/blob/master/design/one-pager-managed-resource-api-design.md#external-resource-name

srueg replied:
Not sure I got this right. Like that?

muvaf replied:
By default, the following initializer is assigned to managed reconciler https://github.com/crossplaneio/crossplane-runtime/blob/master/pkg/resource/managed_reconciler.go#L259

It basically uses `<claim namespace>-<claim-name>-<5 rand characters>` format to create a name and assigns it as an annotation on the managed resource. What you need to do is simply calling `meta.GetExternalName(cr)` like here https://github.com/crossplaneio/stack-aws/blob/master/pkg/controller/cache/managed.go#L182 whenever you need the name of the resource.

srueg replied:
But what if I want to provide the ability to set the name of the external resource in a claim?

muvaf replied:
If your claim has an annotation with key `crossplane.io/external-name`, its value will be used instead of generating a new one.

Nevertheless, you can always use any mechanism you want. This external name propagation mechanism is used across managed resources but if you'd like a field to represent it, that'd also work.

hasheddan replied:
Note: If you do implement your own mechanism you would want to disable the annotation one:

`resource.WithManagedInitializers(resource.NewAPIManagedFinalizerAdder(mgr.GetClient()))`

I recently did this here: https://github.com/packethost/stack-packet/pull/9/files#diff-57be8f6c5cd137378ee9756c8ef40727R66

So that I could set the external name after a device was created in `stack-packet`: https://github.com/packethost/stack-packet/pull/9/files#diff-57be8f6c5cd137378ee9756c8ef40727R178

srueg replied:
Ok, thanks. So I disabled the default external naming and set it on my own (based on the provided value, if any)

negz replied:
> Note: If you do implement your own mechanism you would want to disable the annotation one.

I'm not sure that's true in this case. You need to disable the annotation mechanism for managed resources with non-deterministic external names generated by the external system, to ensure we propagate the externally generated external name back to the annotation.

As @muvaf pointed out, in the default case where you do not disable the external naming functionality by overriding the initializers:

1. Dynamically provisioned managed resources (i.e. those provisioned by a claim) will have their `crossplane.io/external-name` annotation set to the claim's `crossplane.io/external-name` annotation, if any. This happens in the [`ObjectMetaConfigurator`](https://godoc.org/github.com/crossplaneio/crossplane-runtime/pkg/resource#ObjectMetaConfigurator).
1. Statically provisioned managed resources, and dynamically provisioned resources whose claim did not specify an `crossplane.io/external-name` annotation will set that annotation to the name of the managed resource. This happens in the [`ManagedNameAsExternalName`](https://godoc.org/github.com/crossplaneio/crossplane-runtime/pkg/resource#ManagedNameAsExternalName) initializer.

So currently I believe this implementation will:

1. Use the claim's `crossplane.io/external-name` annotation as the bucket name if it is set.
1. Fall back to using a name determined by the `NameFormat` as the bucket name.

I was hoping to deprecate the `NameFormat` pattern - it's an older pattern we used to generate random names for dynamically provisioned managed resources before `ExternalName` gave more control over naming. If you removed the `NameFormat` support and instead used the default `meta.GetExternalName` behaviour the implementation would:

1. Use the claim's `crossplane.io/external-name` annotation as the bucket name if it is set.
1. Fall back to using the managed resource's name, which is `claimnamespace-claimname-randomstring` for dynamically provisioned resources, or whatever name you choose for statically provisioned resources.

Do you have a preference for the older `NameFormat` style?

srueg replied:
Ok, now I got it. I'll change to your proposed behavior regarding `external-name` and remove `NameFormat` style :+1:

hasheddan commented on api/v1alpha1/groupversion_info.go:
```diff
@@ -0,0 +1,70 @@
+var (
+
+	// GroupVersion is group version used to register these objects
+	GroupVersion = schema.GroupVersion{Group: Group, Version: Version}
```
Note: we have traditionally separated out like resources into their own groups (i.e. an S3 bucket would live in the `storage.cloudscale.crossplane.io` group). I do not think this pattern needs to necessarily be followed verbatim, but it may be helpful as you begin to support more resources.

hasheddan reviewed (commented):
Really awesome work here! Left a few comments and tried to link to docs / examples where possible. Please feel free to follow up with any thoughts or questions :)

Commit:
Implement feedback.

* Move S3Bucket types to `storage.cloudscale.crossplane.io` group
* Use external-name annotation for bucket name
* Make optional fields pointer types
* Move provider specific info to `forProvider` and `atProvider` fields

Commit:
Update stack info

Commit:
Fix new group

Commit:
Set observed status during Observe()

Commit:
Disable default external naming behavior

negz commented on controllers/s3/claim.go:
```diff
@@ -0,0 +1,190 @@
+	}
+
+	spec.WriteConnectionSecretToReference = &runtimev1alpha1.SecretReference{
+		Namespace: bucketClaim.Namespace, //s3BucketClass.SpecTemplate.WriteConnectionSecretsToNamespace,
```
It seems like this is ignoring `writeConnectionSecretsToNamespace` and instead using the claim's namespace to store the managed resource's secret. Is this intentional?

srueg replied:
Good catch, thanks. This was actually a test.

negz commented on controllers/s3/s3bucket.go:
```diff
@@ -0,0 +1,265 @@
+	if !ok {
+		return resource.ExternalObservation{}, errors.New(errNotInstance)
+	}
+	log.Info("Observe", "bucket", bucket.Name)
```
I'd be curious to get any thoughts you have on logging - it's pretty light on the ground in Crossplane at the moment but I'm hoping to have a design out soon. My thought is that we'll add a `WithLogger` option to the various controllers that will allow you to pass in a logger to be used, and that said logger would include debug statements for things like this (i.e. calling Observe, Create, etc).

srueg replied:
Yes, that sounds like a reasonable approach. Especially during debugging / testing the logs are a little too sparse and it's hard to see what's going on without a debugger attached.

negz replied:
Do you think these log statements outlining what stage in the reconcile we're at should be at Info or Debug level? My preference would be the latter.

negz commented on controllers/s3/s3bucket.go:
```diff
@@ -0,0 +1,265 @@
+//   2. NameFormat = "%s", BucketName = "test-uid"
+//   3. NameFormat = "foo", BucketName = "foo"
+//   4. NameFormat = "foo-%s", BucketName = "foo-test-uid"
+//   5. NameFormat = "foo-%s-bar-%s", BucketName = "foo-test-uid-bar-%!s(MISSING)"
```
Edit: I see my original comment here duplicates another thread about the external name, so I'm going to discuss it there instead.

negz commented on controllers/suite_test.go:
```diff
@@ -0,0 +1,83 @@
+var k8sClient client.Client
+var testEnv *envtest.Environment
+
+func TestAPIs(t *testing.T) {
```
I'd be curious to get your take on these Ginkgo driven integration style tests vs what we do for our modern controllers (check the tests for any v1beta1 controllers, or the managed resource reconciler itself). How to best test controllers is one area Crossplane and Kubebuilder disagrees with controller-runtime.

srueg replied:
I didn't write tests so far. That would be the next step and I'm happy to give feedback on it.

Commit:
Remove NameFormat from S3BucketSpec

Instead rely on the crossplane.io/external-name annotation on claims and
managed resources [1].

[1] https://github.com/crossplaneio/crossplane/blob/master/design/one-pager-managed-resource-api-design.md#external-resource-name

Commit:
Honor `WriteConnectionSecretsToNamespace` value

Instead of writing the secret to the same namespace as the claim itself.

Commit:
Update CRD printer columns

Commit:
Write bucket name to credentials secret

Commit:
Upgrade to crossplane-runtime v0.2.1

Commit:
Upgrade Cloudscale SDK

Introduce a new field `region` for the bucket type which specifies the
region of a bucket. [1]

Closes #2

[1] https://www.cloudscale.ch/en/news/2019/11/06/geo-redundancy-with-two-cloud-location

negz commented on api/cloudscale.go:
```diff
@@ -0,0 +1,46 @@
+//go:generate controller-gen object:headerFile=../hack/boilerplate.go.txt paths=./...
+
+// Generate crossplane-runtime methodsets (resource.Claim, etc)
+//go:generate angryjet generate-methodsets --header-file=../hack/boilerplate.go.txt ./...
```
https://github.com/crossplaneio/stack-rook/pull/21

You might be interested in the above pattern we arrived on recently to pin `controller-gen` and `angryjet` using Go modules.

srueg replied:
Nice, thanks

negz commented on api/storage/v1alpha1/s3bucket_types.go:
```diff
@@ -0,0 +1,121 @@
+// +kubebuilder:printcolumn:name="CLASS",type="string",JSONPath=".spec.classRef.name"
+// +kubebuilder:printcolumn:name="STATUS",type="string",JSONPath=".status.atProvider.status"
+// +kubebuilder:printcolumn:name="AGE",type="date",JSONPath=".metadata.creationTimestamp"
+// +kubebuilder:resource:scope=Namespaced
```
This one should be cluster scoped, the same as `S3BucketClass`. As of Crossplane 0.4 only resource claims are namespaced.

negz commented on clients/s3/s3.go:
```diff
@@ -0,0 +1,235 @@
+		DisplayName: bucketName,
+		Tags:        bucketTags,
+	}
+	var objectUser *cloudscale.ObjectsUser
```
Our hope is that eventually all Crossplane managed resources will be 1:1 with cloud provider API objects. So in this example, the managed resource would create _only_ a bucket, and not a user. We imagine API objects like this user would be created and managed as a separate managed resource.

That all said, this is totally fine for the moment. If we start modelling (for example) bucket users separately we'll need to arrive on a good pattern for cases like this in which you will almost always want to create a user when you create a bucket. Possibly something like https://github.com/crossplaneio/crossplane/issues/968.

negz commented on clients/s3/s3.go:
```diff
@@ -0,0 +1,235 @@
+		if err != nil {
+			return nil, err
+		}
+	}
```
Totally unrelated to Crossplane, but this if/else block might be easier to follow as a switch statement.

negz commented on controllers/cloudscale.go:
```diff
@@ -0,0 +1,42 @@
+		&s3.BucketClaimSchedulingController{},
+		&s3.BucketClaimDefaultingController{},
+		&s3.BucketClaimController{},
+		&s3.BucketInstanceController{},
```
Nitpick: I'd consider naming this `BucketController`, `BucketResourceController`, or similar. I think there's a little confusion about the term "instance" having meaning in Crossplane, which it does not. We have both claims named instance (e.g. MySQLInstance) and managed resources named instance (e.g. CloudSQLInstance).

negz commented on controllers/s3/claim.go:
```diff
@@ -0,0 +1,184 @@
+	spec.ProviderReference = s3BucketClass.SpecTemplate.ProviderReference
+
+	s3Bucket.Spec = *spec
+	s3Bucket.Namespace = bucketClaim.Namespace
```
This should be removed - managed resources should be cluster scoped.

negz commented on controllers/s3/s3bucket.go:
```diff
@@ -0,0 +1,244 @@
+	s := &corev1.Secret{}
+	n := types.NamespacedName{Namespace: p.Spec.Secret.Namespace, Name: p.Spec.Secret.Name}
+	if err := c.client.Get(ctx, n, s); err != nil {
+		return nil, errors.Wrapf(err, "cannot get Provider secret %v", n.String())
```
Nitpick: I'm pretty sure if you use `%s` instead of `%v` the `.String()` method will be implied.

negz commented on controllers/s3/s3bucket.go:
```diff
@@ -0,0 +1,244 @@
+	// we want to let the resource.ManagedReconciler know so it can create it.
+	if s3.IsErrorNotFound(err) {
+		return resource.ExternalObservation{ResourceExists: false}, nil
+	} else if err != nil {
```
Do you need an `else` here given you return inside the `if` block?

negz commented on controllers/s3/s3bucket.go:
```diff
@@ -0,0 +1,244 @@
+	bucketName := meta.GetExternalName(bucket)
+	objectUser, err := e.s3Client.CreateOrUpdateBucket(ctx, bucket.Status.AtProvider.ObjectUserID, bucketName, bucket.Spec.ForProvider.Region, bucket.Spec.ForProvider.CannedACL, bucket.Spec.ForProvider.Tags)
+	if err != nil {
+		return resource.ExternalCreation{}, errors.Wrap(err, "cannot create instance")
```
Perhaps "cannot create bucket"? We tend to only use the name "instance" when the underlying API object is called an instance, (e.g. an "RDS instance" or "CloudSQL instance").

negz commented on controllers/s3/s3bucket.go:
```diff
@@ -0,0 +1,244 @@
+	}
+
+	bucket.Status.AtProvider.ObjectUserID = objectUser.ID
+	bucket.Status.AtProvider.Status = statusOnline
```
Typically we reserve `AtProvider` for status observations that come from the actual provider (i.e. that are pretty much direct translations of the cloud provider API). This status seems to be a concept introduced by this controller - perhaps it should be directly under `bucket.Status`?

negz commented on controllers/s3/s3bucket.go:
```diff
@@ -0,0 +1,244 @@
+// Create a new external resource based on the specification of our managed
+// resource. resource.ManagedReconciler only calls Create if Observe reported
+// that the external resource did not exist.
+func (e *external) Create(ctx context.Context, mg resource.Managed) (resource.ExternalCreation, error) {
```
It's typical to set the creating status condition here; the condition represents our controller's intent to create the bucket as well as the actual observation that it is creating (deleting, etc).

negz reviewed (changes requested):
A few comments, but the only real concern is that the bucket managed resource is namespaced. This would break convention with all other Crossplane managed resources, which are cluster scoped.

Commit:
Pin code gen tooling

Commit:
Make S3Bucket cluster scoped

Commit:
Implement various review findings

Commit:
Bump version to v0.0.2

srueg commented:
Thanks y'all for the feedback, really appreciate it!

negz reviewed (approved):
This looks great! Thanks for bearing with us - we're conscious that we've been moving pretty fast with our patterns and APIs.

srueg commented:
No problem and also thanks for bearing with me, I'm still getting into Golang and your feedback is great!

tobru commented:
This is just great. Thanks to all involved in review and for the feedback.
